File: leetcode/1008-construct-binary-search-tree-from-preorder-traversal/solution.cpp
#include <climits>
#include <vector>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        i = 0;
        this->preorder = &preorder;
        return build(LONG_MAX);
    }

private:
    vector<int>* preorder;
    int i;

    TreeNode* build(long bound) {
        if (i == (int)preorder->size() || (*preorder)[i] > bound) {
            return nullptr;
        }
        TreeNode* node = new TreeNode((*preorder)[i]);
        i++;
        node->left = build(node->val);
        node->right = build(bound);
        return node;
    }
};

File: leetcode/1008-construct-binary-search-tree-from-preorder-traversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int[] preorder;
    private int i;

    public TreeNode bstFromPreorder(int[] preorder) {
        this.preorder = preorder;
        this.i = 0;
        return build(Long.MAX_VALUE);
    }

    private TreeNode build(long bound) {
        if (i == preorder.length || preorder[i] > bound) {
            return null;
        }
        TreeNode node = new TreeNode(preorder[i]);
        i++;
        node.left = build(node.val);
        node.right = build(bound);
        return node;
    }
}
